netket/optimizer/qgt/default.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, Optional, Union, Tuple, Any
+from functools import partial
+
+import jax
+from jax import numpy as jnp
+from flax import struct
+
+from netket.utils.types import PyTree
+from netket.utils import n_nodes
+from netket.stats import sum_inplace
+import netket.jax as nkjax
+
+from .qgt_jacobian import QGTJacobian
+from .qgt_onthefly import QGTOnTheFly
+
+
+def default_qgt_matrix(variational_state, solver=False):
+    # arbitrary heuristic: for more than 2
+
+    # an rbm has 3
+    n_param_blocks = len(jax.tree_leaves(variational_state.parameters))
+    n_params = variational_state.n_params
+
+    # Completely arbitrary
+    if n_param_blocks > 6 and n_params > 800:
+        return QGTJacobianPyTree
+    else:
+        return QGTOnTheFly
```

Comment:
This attempts to be smart about when to switch from `QGTOnTheFly` to `QGTJacobianPyTree` (still unimplemented, better name?). 
however this auto switch should be benchmarked a bit I think..

Reply:
As it is now, `QGTOnTheFly` is the only one that handles most parameter/result dtypes correctly. If there's interest, I can start a PR on `QGTJacobianPyTree` that implements https://github.com/netket/netket/pull/652#issuecomment-827949805, so it would be competitive.

Reply:
Names: `QGTJacobianPyTree` is good, but we should change the matrix-based one to `QGTJacobianMatrix` or something and alias our favourite (probs the PyTree-based one once it's written) as `QGTJacobian`.

Reply:
Heuristics: the main bottleneck for onthefly is backpropagating through several layers (i.e. a lot of nonlinear steps) rather than linear algebra in any of them, so `n_param_blocks` will probably be the one we want to check. Although, you end up with loads of empty leaves and I don't know whether to worry about those...

Reply:
`QGTJacobianMatrix` -> `QGTJacobianDense`.
Seems good.
Do you get empty leaves, really? I never saw this...

Reply:
Leaves should only be jax arrays by definition.
Do you have an example where this happens?

Reply:
IIRC Stax machines (which I mostly use) produce an empty leaf for every layer without parameters (like activations). But I don't swear by it, I'll go back and check

Reply:
I think they have empty nodes, but those are not leaf nodes. Otherwise `jax.tree_map` wouldn't work correctly on them (and NetKet wouldn't work, because it uses tons of `jax.tree_map`)
But I might be wrong, so I'd love a counterexample if you have one.

Reply:
>  If there's interest, I can start a PR on QGTJacobianPyTree that implements #652 (comment), so it would be competitive.

Feel free to do so. There's a bunch of code already there, we just need to properly hook it up.
If you open a pr open it against this branch